### PR TITLE
Docker update

### PR DESCRIPTION
**Pull request by spficklin**
This PR updates the Docker file to support the RScripts and the 3D python viewer so that users who want to use the Docker image to avoid installing all of the dependencies can do so.  

This PR makes the following changes:
- In the Makefile:
  - the `lastest` images now have ACE hardcoded to 3.2.0 rather than `develop`. This is because the develop branch can have, at any point in time, experimental code.  I think the `latest` image should be the latest of development version of KINC and not the latest of of ACE.
  - I changed the NVIDIA_HEADLESS option to be consistent between the cpu and gpu tagged images.
  - I added a new KINC_R_REVISION argument.
- In the Dockerfile
  - I updated section to behave differently for newer version of KINC. For example in 3.4.2 we have a much simplified make procedure that if not used does not install the scripts in the bin folder.
  - I added installation of R, KINC.R
- In the Documentation 
  - I adjust the text to reflect the new additions for the docker images. However, I'm not 100% sure when to tell the user to use the `-cpu` and `-gpu` tagged images.  That should be included as the current image name isn't correct in the commands.
   - I changed the docker command from `nvidia-docker` to just `docker` with the `--gpus all` option.  I wasn't able to run the image with `nvidia-docker` as that didn't exist on my machine.

Unfortunately, all of these changes makes the image rather large and take a long time to build.

### CODE REVIEW

**Comment by spficklin** [comment]
@bentsherman would you mind looking this over?  If you don't have time for a functional review if you can just check the changes and see if you are okay with them?  

The only outstanding item on this is clarification about the `-cpu` and `-gpu` tags on the images which is missing in the documentation.

**Comment by spficklin** [comment]
Also, once this PR is merged and after we do the 3.4.2 release we can rebuild the images for DockerHub.  For the 3.4.2 release we just have this PR and the other #171 .  Then we can release!

**Comment by spficklin** [comment]
Oh, one other note.... If you try to build the images you'll see some warnings when building KINC about unused variables. That has been fixed in #171 so no need to try to clean that up. It will go away once all the PRs are merged.

File: docker/Dockerfile [hunk 1]
```diff
@@ -43,11 +45,10 @@ RUN git clone -q https://github.com/kthohr/stats \
 # install ACE
 WORKDIR /opt
 
-ADD https://api.github.com/repos/SystemsGenetics/ACE/git/refs/heads/${ACE_REVISION} ace-version.json
```
Comment (bentsherman) [inline docker/Dockerfile:46]: Why was this line removed? It allows you to push commits to ACE and then rebuild the docker image with the new commits. Otherwise docker would cache the previous git clone command and not pull the new commits. ALso the line below for cloning the KINC repo.

File: docker/Dockerfile [hunk 2]
```diff
@@ -43,11 +45,10 @@ RUN git clone -q https://github.com/kthohr/stats \
...
 RUN git clone -q https://github.com/SystemsGenetics/ACE.git \
 	&& cd ACE/build \
 	&& git checkout -q ${ACE_REVISION} \
-	&& qmake ../src/ACE.pro GUI=no \
```
Comment (bentsherman) [inline docker/Dockerfile:50]: Why was this setting removed? I believe this setting helped decrease the size of the docker image. I don't think you can use the GUI through a docker image anyway.

File: docker/Makefile [hunk 3]
```diff
@@ -3,41 +3,45 @@ all: latest v3.4.x v3.3.x v3.2.x
 latest:
 	docker build \
 		-t systemsgenetics/kinc:latest-cpu \
-		--build-arg NVIDIA_HEADLESS=1 \
-		--build-arg ACE_REVISION=develop \
+		--build-arg NVIDIA_HEADLESS=0 \
```
Comment (bentsherman) [inline docker/Makefile]: The NVIDIA_HEADLESS settings are backwards now. It should be enabled for CPU build and disabled for GPU build. nvidia-headless is basically a stub for the actual nvidia drivers. Because of the way KINC works, you have to have the NVIDIA drivers installed on your system or else KINC will not run. So for CPU-only machines (which would use the CPU docker image), you would install the nvidia-headless drivers to keep KINC happy without having to install the actual NVIDIA drivers. This trick is also how you would run KINC natively on a CPU-only machine.

File: docker/Makefile [hunk 4]
```diff
@@ -3,41 +3,45 @@ all: latest v3.4.x v3.3.x v3.2.x
...
+		--build-arg ACE_REVISION=v3.2.0 \
 		--build-arg KINC_REVISION=develop \
+		--build-arg KINC_R_REVISION=v1.2 \
 		.
 
 v3.4.x:
 	docker build \
 		-t systemsgenetics/kinc:3.4.2-cpu \
-		--build-arg NVIDIA_HEADLESS=1 \
+		--build-arg NVIDIA_HEADLESS=0 \
```
Comment (bentsherman) [inline docker/Makefile]: NVIDIA_HEADLESS is backwards

File: docker/Makefile [hunk 5]
```diff
@@ -3,41 +3,45 @@ all: latest v3.4.x v3.3.x v3.2.x
...
 		--build-arg ACE_REVISION=v3.2.0 \
 		--build-arg KINC_REVISION=v3.4.2 \
+		--build-arg KINC_R_REVISION=v1.2 \
 		.
 
 v3.3.x:
 	docker build \
 		-t systemsgenetics/kinc:3.3.0-cpu \
-		--build-arg NVIDIA_HEADLESS=1 \
+		--build-arg NVIDIA_HEADLESS=0 \
```
Comment (bentsherman) [inline docker/Makefile]: NVIDIA_HEADLESS is backwards

File: docker/Dockerfile [hunk 6]
```diff
@@ -57,20 +58,72 @@ ENV LD_LIBRARY_PATH "/usr/local/lib:$LD_LIBRARY_PATH"
...
-	&& mkdir -p KINC/build \
-	&& cd KINC/build \
-	&& git checkout -q ${KINC_REVISION} \
-	&& qmake ../src/KINC.pro GUI=no \
-	&& make -s -j $(nproc) \
-	&& make -s qmake_all \
-	&& make -s install
+# For older version of KINC we need to use qmake. For newer versions after 3.3.x we
+# can use the make file in the root directory of KINC.
+RUN if [ "${KINC_REVISION}" = "v3.3.x" ] || [ "${KINC_REVISION}" = "v3.2.x" ] ; then \
```
Comment (bentsherman) [inline docker/Dockerfile]: I don't think these conditionals will work because KINC_REVISION must be set to a specific version like `v3.3.0` rather than `v3.3.x`, so the docker build might end up running all of these conditionals. You need to check your docker build log and make sure it executed the commands you expected.

File: docker/Dockerfile [hunk 7]
```diff
@@ -57,20 +58,72 @@ ENV LD_LIBRARY_PATH "/usr/local/lib:$LD_LIBRARY_PATH"
...
+        git clone -q https://github.com/SystemsGenetics/KINC.git \
+        && cd KINC/build \
+    	&& git checkout-q ${KINC_REVISION} \
```
Comment (bentsherman) [inline docker/Dockerfile]: Missing a space between checkout and -q

File: docker/Dockerfile (hunk 6 above)
Comment (bentsherman) [reply docker/Dockerfile]: Perhaps a better way would be to check for the existence of Makefile, because I added the Makefile at the same time I added the R scripts to the build.

File: docker/Dockerfile [hunk 8]
```diff
@@ -57,20 +58,72 @@ ENV LD_LIBRARY_PATH "/usr/local/lib:$LD_LIBRARY_PATH"
...
-RUN pip3 install -q -r requirements.txt
+RUN pip3 install  -r requirements.txt
+
+
 
 # initialize default work directory
 WORKDIR /workspace
+
+# Tini for signal processing and zombie killing.
+ENV TINI_VERSION v0.18.0
```
Comment (bentsherman) [inline docker/Dockerfile]: Tini is no longer needed in docker builds, refer to Tini github repo: https://github.com/krallin/tini#using-tini

**Review by bentsherman — CHANGES_REQUESTED** [review]
I left my review via inline comments. Just a few things that need to be changed. The docker image will likely be a lot bigger because of the additional R dependencies, but hopefully that will all be removed once the R scripts are migrated to Python. A few extra notes:
- Make sure that you can actually use the KINC GUI from the docker image. I know you can run web servers from a docker image but I have never tried native GUI applications. If KINC GUI doesn't work then you should leave the `GUI=no` option in the build and specify that in the documentation.
- The CPU builds are for CPU-only machines, and it installs the NVIDIA headless drivers instead of the actual drivers so that KINC loads correctly. You may recall users in the past reporting issues with running KINC without a GPU, and this is the workaround that I developed for that.

File: docker/Dockerfile (hunk 1 above)
Comment (spficklin) [reply docker/Dockerfile:46]: It doesn't work for tagged releases, which makes sense as those don't change.  I can try and wrap it in an if statement in case the revision is 'develop'.

File: docker/Dockerfile (hunk 2 above)
Comment (spficklin) [reply docker/Dockerfile:50]: Probably just my mistake. I can add it back in.

File: docker/Makefile (hunk 3 above)
Comment (spficklin) [reply docker/Makefile]: Oh, why is called 'headless'?   Shouldn't it 'gpuless'?  But I'll switch it back.

File: docker/Dockerfile (hunk 6 above)
Comment (spficklin) [reply docker/Dockerfile]: Yes, good catch. That was my stupid mistake.  I'll fix it.

File: docker/Dockerfile (hunk 8 above)
Comment (spficklin) [reply docker/Dockerfile]: I included it because I could not ctrl+c out of a docker image once it was running. After I added it I was able to do so.  I didn't realize I could just pass the `--init` flag to `docker run`. 
I don't see any harm in leaving it so we don't have to require the flag.

File: docker/Dockerfile (hunk 1 above)
Comment (spficklin) [reply docker/Dockerfile:46]: It seems you can't use an IF statement for an ADD statement.  I'm not sure how to fix this one.

File: docker/Dockerfile (hunk 2 above)
Comment (spficklin) [reply docker/Dockerfile:50]: Fixed in latest push to the branch

File: docker/Makefile (hunk 3 above)
Comment (spficklin) [reply docker/Makefile]: fixed in latest push to the branch

File: docker/Makefile (hunk 5 above)
Comment (spficklin) [reply docker/Makefile]: fixed in latest push to the branch

File: docker/Dockerfile (hunk 1 above)
Comment (bentsherman) [reply docker/Dockerfile:46]: Okay now I remember for tagged releases you have to use the URL `refs/tags` instead of `refs/heads`. So you can add a conditional for that, or just leave it as is if you don't want to deal with it. Not a critical feature, just makes things easier if you only updated the github repo before a docker build.

File: docker/Dockerfile (hunk 7 above)
Comment (spficklin) [reply docker/Dockerfile]: Good catch. I forgot to test a v3.3.x build, which is why I didn't catch these. I'll make sure I build one of these..  

This has been fixed in the latest push to the branch

File: docker/Makefile (hunk 4 above)
Comment (spficklin) [reply docker/Makefile]: fixed in latest push to the branch

File: docker/Dockerfile (hunk 1 above)
Comment (spficklin) [reply docker/Dockerfile:46]: I don't think it's possible to add a conditional for ADD statements :-(  But I'm happy to do it if you know how. I couldn't find a way to do it.

File: docker/Dockerfile (hunk 6 above)
Comment (spficklin) [reply docker/Dockerfile]: This has been fixed in the most recent push.

**Comment by spficklin** [comment]
Hi @bentsherman. I believe I have resolved all of the issues you raised on this PR.  I did update the Dockerfile a bit more to make it easier to build just a cpu or gpu version.  Also I added a new `develop` tag and set `latest` to point to the actual latest release rather than the develop branch.  Other than that, the only other changes I made were to fix issues.  

Can you review again?  I'm hoping to do a release on Monday for Iris.

**Review by bentsherman — APPROVED** [review]

**Comment by spficklin** [comment]
Thanks @bentsherman
